nixos/modules/services/desktops/pipewire.nix

```diff
@@ -46,32 +98,112 @@ in {
+            "api.vulkan.*" = "vulkan/libspa-vulkan";
+            "api.jack.*" = "jack/libspa-jack";
+            "support.*" = "support/libspa-support";
+            # "videotestsrc" = "videotestsrc/libspa-videotestsrc";
+            # "audiotestsrc" = "audiotestsrc/libspa-audiotestsrc";
+          };
+
+          modules = {
+            ##  <module-name> = { [args = "<key>=<value> ..."]
+            #                     [flags = ifexists] }
+            #                     [flags = [ifexists]|[nofail]}
+            #
+            # Loads a module with the given parameters.
+            # If ifexists is given, the module is ignoed when it is not found.
+            # If nofail is given, module initialization failures are ignored.
+            #
+            libpipewire-module-rtkit = {
+              args = "\"rt.prio=20 rt.time.soft=200000 rt.time.hard=200000\"";
+              flags = "ifexists|nofail";
+            };
+            libpipewire-module-protocol-native = { _priority = -100; _content = "null"; };
+            libpipewire-module-profiler = "null";
+            libpipewire-module-metadata = "null";
+            libpipewire-module-spa-device-factory = "null";
+            libpipewire-module-spa-node-factory = "null";
+            libpipewire-module-client-node = "null";
+            libpipewire-module-client-device = "null";
+            libpipewire-module-portal = "null";
+            libpipewire-module-access = {
+              args = "\"access.allowed=${builtins.unsafeDiscardStringContext cfg.sessionManager} access.force=flatpak\"";
+            };
+            libpipewire-module-adapter = "null";
+            libpipewire-module-link-factory = "null";
+            libpipewire-module-session-manager = "null";
+          };
+
+          objects = {
+            ## create-object [-nofail] <factory-name> [<key>=<value> ...]
+            #
+            # Creates an object from a PipeWire factory with the given parameters.
+            # If -nofail is given, errors are ignored (and no object is created)
+            #
+          };
+
+
+          exec = {
+            ## exec <program-name>
+            #
+            # Execute the given program. This is usually used to start the
+            # session manager. run the session manager with -h for options
+            #
+            "${builtins.unsafeDiscardStringContext cfg.sessionManager}" = { args = "\"\""; };
+          };
+        };
       };
 
       sessionManager = mkOption {
-        type = types.nullOr types.string;
-        default = null;
-        example = literalExample ''"''${pipewire}/bin/pipewire-media-session"'';
+        type = types.str;
+        default = "${cfg.package}/bin/pipewire-media-session";
+        example = literalExample ''"\$\{pipewire\}/bin/pipewire-media-session"'';
         description = ''
           Path to the pipewire session manager executable.
         '';
       };
 
-      sessionManagerArguments = mkOption {
```

Comment:
I was thinking of adding an extraEtcFiles instead since that's the way to configure pipewire-media-session now. But for other session managers it makes sense to keep this around I guess.

Reply:
maybe `settings` as per https://github.com/NixOS/rfcs/pull/42?

Reply:
The problem is that depending on the session manager there can be an arbitrary number of config files and their names aren't known ahead of time. The current solution allows specifying arbitrary files and if the `toSPAJSON` function is made public (just looking for the right place for it) it would allow specifying settings for `pipewire-media-session` mostly in this way, just like is done with pipewire.conf (except for having to manually call the fn). But a fancy config generation setup can be done separately too.